Cabal/Distribution/Simple/Configure.hs

```diff
@@ -1397,13 +1397,18 @@ combinedConstraints constraints dependencies installedPackages = do
     -- (i.e. someone has written a hash) and didn't find it then it's
     -- an error.
     badComponentIds =
-      [ (pkgname, cid)
-      | (pkgname, cid, Nothing) <- dependenciesPkgInfo ]
+      [ (pkgname, cname, cid)
+      | (pkgname, cname, cid, Nothing) <- dependenciesPkgInfo ]
 
     dispDependencies deps =
       hsep [    text "--dependency="
-             <<>> quotes (disp pkgname <<>> char '=' <<>> disp cid)
-           | (pkgname, cid) <- deps ]
+             <<>> quotes
+                    (disp pkgname
+                    <<>> case cname of CSubLibName sublib -> ":" <<>> disp sublib
+                                       _ -> "" --XXX
```

Comment:
(A) I'm not sure about this. Maybe I should put error "impossible non-library dependecy" if I get something different than CLibName or CSubLibName. Or maybe i should add an entirely different datatype (probably not worth it).

Reply:
In this case, since this is "display" code, you should still print out the non-library component in the "obvious way". If you just return an empty "" or put an error it will make it more difficult to debug if you end up with something invalid here.